### PR TITLE
KUBESAW-12: Convert the health-check goroutine into ToolchainCluster controller

### PR DESCRIPTION
**Pull request by fbm3307**
This is to convert the health-check goroutine into ToolchainCluster controller

The related PRs:

- Toolchain-Common - https://github.com/codeready-toolchain/toolchain-common/pull/386
- Member-Operator - https://github.com/codeready-toolchain/member-operator/pull/559

The Other PRs where only Toolchain common dependency is updated 
- Toolchain-e2e - https://github.com/codeready-toolchain/toolchain-e2e/pull/949
- Registration-service - https://github.com/codeready-toolchain/registration-service/pull/422
- Sandbox-sre - https://github.com/codeready-toolchain/sandbox-sre/pull/1698

### CODE REVIEW

**Comment by rsoaresd** [comment]
Hello @fbm3307,

The last run ([1780159346131341312](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/codeready-toolchain_host-operator/1010/pull-ci-codeready-toolchain-host-operator-master-e2e/1780159346131341312)) failed with:

```
=== RUN   TestToolchainClusterE2E/create_new_ToolchainCluster_based_on_'host'_with_incorrect_data_and_expect_to_be_offline
    awaitility.go:539: waiting for toolchaincluster in namespace 'toolchain-host-16094520' to match criteria
    toolchaincluster_test.go:113: 
        	Error Trace:	/tmp/toolchain-e2e/test/e2e/toolchaincluster_test.go:113
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestToolchainClusterE2E/create_new_ToolchainCluster_based_on_'host'_with_incorrect_data_and_expect_to_be_offline
--- FAIL: TestToolchainClusterE2E (124.34s)
    --- PASS: TestToolchainClusterE2E/create_new_ToolchainCluster_based_on_'host'_with_correct_data_and_expect_to_be_ready (2.94s)
    --- FAIL: TestToolchainClusterE2E/create_new_ToolchainCluster_based_on_'host'_with_incorrect_data_and_expect_to_be_offline (121.39s)
```

When looking at [this PR history](https://prow.ci.openshift.org/pr-history/?org=codeready-toolchain&repo=host-operator&pr=1010), we can observe that [1778725233843245056](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/codeready-toolchain_host-operator/1010/pull-ci-codeready-toolchain-host-operator-master-e2e/1778725233843245056) failed with the same error. The other runs related to this PR failed due to infra issues.


From [this search query](https://search.dptools.openshift.org/?search=%2Ftmp%2Ftoolchain-e2e%2Ftest%2Fe2e%2Ftoolchaincluster_test.go%3A113&maxAge=336h&context=1&type=build-log&name=.*codeready-toolchain.*&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job), we can see that happens only in this PR. So, very likely is PR specific. Hope it helps :)

**Review by mfrancisc — APPROVED** [review]

**Comment by fbm3307** [comment]
> Hello @fbm3307,
> 
> The last run ([1780159346131341312](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/codeready-toolchain_host-operator/1010/pull-ci-codeready-toolchain-host-operator-master-e2e/1780159346131341312)) failed with:
> 
> ```
> === RUN   TestToolchainClusterE2E/create_new_ToolchainCluster_based_on_'host'_with_incorrect_data_and_expect_to_be_offline
>     awaitility.go:539: waiting for toolchaincluster in namespace 'toolchain-host-16094520' to match criteria
>     toolchaincluster_test.go:113: 
>         	Error Trace:	/tmp/toolchain-e2e/test/e2e/toolchaincluster_test.go:113
>         	Error:      	Received unexpected error:
>         	            	timed out waiting for the condition
>         	Test:       	TestToolchainClusterE2E/create_new_ToolchainCluster_based_on_'host'_with_incorrect_data_and_expect_to_be_offline
> --- FAIL: TestToolchainClusterE2E (124.34s)
>     --- PASS: TestToolchainClusterE2E/create_new_ToolchainCluster_based_on_'host'_with_correct_data_and_expect_to_be_ready (2.94s)
>     --- FAIL: TestToolchainClusterE2E/create_new_ToolchainCluster_based_on_'host'_with_incorrect_data_and_expect_to_be_offline (121.39s)
> ```
> 
> When looking at [this PR history](https://prow.ci.openshift.org/pr-history/?org=codeready-toolchain&repo=host-operator&pr=1010), we can observe that [1778725233843245056](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/codeready-toolchain_host-operator/1010/pull-ci-codeready-toolchain-host-operator-master-e2e/1778725233843245056) failed with the same error. The other runs related to this PR failed due to infra issues.
> 
> From [this search query](https://search.dptools.openshift.org/?search=%2Ftmp%2Ftoolchain-e2e%2Ftest%2Fe2e%2Ftoolchaincluster_test.go%3A113&maxAge=336h&context=1&type=build-log&name=.*codeready-toolchain.*&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job), we can see that happens only in this PR. So, very likely is PR specific. Hope it helps :)

Hey @rsoaresd Thanks for looking into this, but it passed this time, i dont think it was related to my PR, but it was flaky

**Comment by MatousJobanek** [comment]
> Hey @rsoaresd Thanks for looking into this, but it passed this time, i dont think it was related to my PR, but it was flaky

TBH, I haven't seen this error (failed test) ever before. I doubt that it would be an old flaky test. This error seems to be rather caused by the changes in the PR.

File: main.go [hunk 1]
```diff
@@ -223,6 +225,15 @@ func main() { // nolint:gocyclo
 		setupLog.Error(err, "unable to create controller", "controller", "ToolchainClusterCache")
 		os.Exit(1)
 	}
+
+	if err := toolchainclusterhealth.NewReconciler(
+		mgr,
+		requeAfter,
+	).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "ToolchainClusterHealth")
```
Comment (MatousJobanek) [inline main.go:233]:
```suggestion
		setupLog.Error(err, "unable to create controller", "controller", "ToolchainCluster")
```

File: main.go [hunk 2]
```diff
@@ -223,6 +225,15 @@ func main() { // nolint:gocyclo
 		setupLog.Error(err, "unable to create controller", "controller", "ToolchainClusterCache")
 		os.Exit(1)
 	}
+
+	if err := toolchainclusterhealth.NewReconciler(
+		mgr,
+		requeAfter,
```
Comment (MatousJobanek) [inline main.go:231]: let's either rename the variable to make it clear that it's related to `ToolchainCluster` reconciliation, or pass it directly to the function - `requeAfter` is pretty generic name
```suggestion
		10 * time.Second,
```

**Review by MatousJobanek — COMMENTED** [review]
Could you please update the dependency to point to the latest version in your toolchain-common PR?

File: main.go (hunk 1 above)
Comment (fbm3307) [reply main.go:233]: done

File: main.go (hunk 2 above)
Comment (fbm3307) [reply main.go:231]: done

File: main.go [hunk 3]
```diff
@@ -223,6 +224,16 @@ func main() { // nolint:gocyclo
 		setupLog.Error(err, "unable to create controller", "controller", "ToolchainClusterCache")
 		os.Exit(1)
 	}
+
+	if err := (&toolchaincluster.Reconciler{
+		Client:     mgr.GetClient(),
+		Scheme:     mgr.GetScheme(),
+		RequeAfter: 10 * time.Second,
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "ToolchainClusterHealth")
```
Comment (rajivnathan) [inline main.go:233]:
```suggestion
		setupLog.Error(err, "unable to create controller", "controller", "ToolchainCluster")
```
I still see it as 'ToolchainClusterHealth' in the PR, maybe the change got reverted somehow?

Comment (fbm3307) [reply main.go]: done

**Review by rajivnathan — APPROVED** [review]
👍

**Pull request merged**